cpp/src/arrow/types/construct.cc

```diff
@@ -71,6 +72,20 @@ Status MakeBuilder(MemoryPool* pool, const std::shared_ptr<DataType>& type,
   }
 }
 
+Status MakeStructBuilder(MemoryPool* pool, const std::shared_ptr<DataType>& type,
```

Comment:
It would be nice to avoid requiring the type to be passed in and constructing the DataType based on the fields passed in.  

Also, we can take another look when we have start using this code more but it might be nice to pass in a vector of builders instead of there data-types/fields.